osquery/TPipe.py

```diff
@@ -0,0 +1,204 @@
+            try:
+                win32pipe.WaitNamedPipe(self._pipeName, self._timeout)
+            except Exception as e:
+                if e.args[0] not in (winerror.ERROR_SEM_TIMEOUT,
+                                     winerror.ERROR_PIPE_BUSY):
+                    raise TTransportException(
+                        type=TTransportException.UNKNOWN,
+                        message='Client failed to connect to server with {}'.
+                        format(e.args[0]))
+        self._handle = h
+
+    def read(self, sz):
+        if not self.isOpen():
+            raise TTransportException(
+                type=TTransportException.NOT_OPEN,
+                message='Called read on non-open pipe')
+        buff = None
+        err = None
+        try:
+            (err, buff) = win32file.ReadFile(self._handle, sz, None)
+        except Exception as e:
+            raise TTransportException(
+                type=TTransportException.UNKNOWN, message='TPipe read failed')
+
+        if (err != 0):
+            raise TTransportException(
+                type=TTransportException.UNKNOWN,
+                message='TPipe read failed with GLE={}'.format(err))
+        if len(buff) == 0:
+            raise TTransportException(
+                type=TTransportException.END_OF_FILE,
+                message='TPipe read 0 bytes')
+        return buff
+
+    def write(self, buff):
+        if not self.isOpen():
+            raise TTransportException(
+                type=TTransportException.NOT_OPEN,
+                message='Called read on non-open pipe')
+
+        bytesWritten = None
+        try:
+            (writeError, bytesWritten) = win32file.WriteFile(
+                self._handle, buff, None)
+        except Exception as e:
+            raise TTransportException(
+                type=TTransportException.UNKNOWN,
+                message='Failed to write to named pipe: ' + e.message)
+
+    def flush(self):
+        win32file.FlushFileBuffers(self._handle)
+
+
+class TPipeServer(TPipeBase, TServerTransportBase):
+    """Named pipe implementation of TServerTransport"""
+
+    def __init__(self, pipeName, buffsize=4096, maxconns=255):
+        """
+        Initialize a TPipe client
+
+        @param pipeName(string)  The path of the pipe to connect to.
+        @param buffsize(int)  The size of read/write buffers to use.
+        @param maxconns(int)  Maximum number of simultaneous connections.
+        """
+        self._pipeName = pipeName
+        self._buffsize = buffsize
+        self._maxconns = maxconns
+        self._handle = None
+
+        # Overlapped event for Windows Async IO
+        self._overlapped = pywintypes.OVERLAPPED()
+        self._overlapped.hEvent = win32event.CreateEvent(None, 0, 0, None)
+
+    # Create a new named pipe if one doesn't already exist
+    def listen(self):
+        if self._handle == None:
+            ret = self.createNamedPipe()
+
+        if ret != True:
```

Comment:
If `self._handle is not None` `ret` will be undefined here. `createNamedPipe()` should also throw on error so `ret` is unnecessary.